Fixing index out of range during exec of container

rajasec opened the pull request:
During exec of container, panic thrown for index out of range
exec failed: panic from initialization: runtime error: index out of range, /home/raj/go/pkg/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/opencontainers/runc/libcontainer/factory_linux.go:255 (0x4b763f)
/usr/local/go/src/runtime/asm_amd64.s:401 (0x43b525)
/usr/local/go/src/runtime/panic.go:387 (0x428ee8)
/usr/local/go/src/runtime/panic.go:12 (0x42807e)
/home/raj/go/pkg/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/opencontainers/runc/libcontainer/setns_init_linux.go:52 (0x4b44e8)
/home/raj/go/pkg/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/opencontainers/runc/libcontainer/factory_linux.go:263 (0x49fd04)
/home/raj/go/pkg/src/github.com/opencontainers/runc/start.go:104 (0x410fbc)
/home/raj/go/pkg/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/codegangsta/cli/command.go:137 (0x482213)
/home/raj/go/pkg/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/codegangsta/cli/app.go:176 (0x47ef35)
/home/raj/go/pkg/src/github.com/opencontainers/runc/main.go:123 (0x40763f)
/usr/local/go/src/runtime/proc.go:63 (0x42aa33)
/usr/local/go/src/runtime/asm_amd64.s:2232 (0x43d641)

Added the check so that process arguments can not be empty during exec.

Signed-off-by: rajasec <rajasec79@gmail.com>

marcosnils commented:
+1.

crosbymichael commented on exec.go:
```diff
@@ -142,6 +142,9 @@ func getProcess(context *cli.Context, bundle string) (*specs.Process, error) {
 	}
 	p := spec.Process
 	p.Args = context.Args()[1:]
+	if len(p.Args) == 0 {
```
Shouldn't this be done before we get `context.Args()`?

rajasec commented:
@crosbymichael 
I have moved up the check. Also I've added for empty path check, because if "-p" option is not specified and there is no arguments specified in cli flags, throw out the error.

marcosnils commented on exec.go:
```diff
@@ -99,6 +99,10 @@ func execProcess(context *cli.Context) (int, error) {
 	if err != nil {
 		return -1, err
 	}
+	path := context.String("process")
```
What if I do `runc exec -p  <container_id> ? 

Wouldn't https://github.com/opencontainers/runc/pull/740/files#diff-cc79fe39318bd23b3e2242c0aaac24fdR149 throw index out of range also?

rajasec commented:
@marcosnils 
./runc exec -p 
incorrect usage
flag needs an argument: -p

./runc exec test
process args cannot be empty

./runc exec
exec failed: container id cannot be empty

marcosnils commented:
@rajasec yeah, I've realized aboutt his afterwards, that's why I deleted my comment :wink:

rajasec commented:
@crosbymichael 
Do I need to make any further changes ?

crosbymichael commented:
LGTM

cyphar commented on exec.go:
```diff
@@ -99,6 +99,10 @@ func execProcess(context *cli.Context) (int, error) {
 		return -1, err
 	}
+	path := context.String("process")
+	if path == "" && len(context.Args()[1:]) == 0 {
```
Why not just `len(context.Args()) == 1`?

rajasec commented:
when len is ==1 no issues. the problem occured when there is no arguments. so i'm comparing when len is 0. if we add <=1, then it will fail even if there is one argument  as failure which is not the case.

cyphar commented:
@rajasec Okay, then do `len(context.Args()) == 1`. The slicing looks odd IMO.

rajasec commented:
@cyphar 
Updated as per comments.

cyphar commented:
LGTM.

hqhq commented:
LGTM
